### PR TITLE
[JENKINS-21224] Cannot use `Listeners.notify` in `ItemListener.fireOnCreated` due to `matrix-auth` tricks

### PR DESCRIPTION
**Pull request by jglick**
https://github.com/jenkinsci/jenkins/pull/6000#discussion_r766087688

### Proposed changelog entries

* Newly created items are again automatically made accessible to their creators to fix a regression in the `matrix-auth` plugin (regression in 2.324).

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by basil — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by daniel-beck** [comment]
I'd be happy to change `matrix-auth` if you have a suggestion how to accomplish the same differently.

**Comment by timja** [comment]
test failure might be related?

File: core/src/main/java/hudson/model/listeners/ItemListener.java
Comment (basil) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:208]:
```suggestion
```

Comment (basil) [reply core/src/main/java/hudson/model/listeners/ItemListener.java:208]: I think this should be deleted now that we are restoring the old version of the code.

**Review by basil — APPROVED** [review]

**Comment by daniel-beck** [comment]
> I do not. It is just generally expected that events are delivered by a system thread carrying no authentication.

That doesn't really make sense as it's trivial to get from any authentication to `SYSTEM`, but not trivial to do the opposite, so this loses information about who triggered an event. Should we have an `Authentication` holder for the request somewhere, to accomplish that?

**Comment by jglick** [comment]
> This loses information about who triggered an event

Indeed, but the normal usage of listeners is that they simply react to a change somewhere, but the listener itself is trusted (`SYSTEM`) not something done “by” a user.

> Should we have an `Authentication` holder for the request somewhere

I guess for listeners pertaining to user config modifications it would make sense for the listener methods to take an `Authentication` argument.

> Does it now attribute all config changes to `SYSTEM`?

Possibly. I think it uses `SaveableListener`. Not sure offhand.

**Comment by jglick** [comment]
I took a look at `jobConfigHistory` and it uses a variety of listeners and does capture current auth in a number of cases (not always clear whether this information is used), so to be on the safe side I tried to classify existing `notify` calls into ones that should or should not pass along auth.

File: core/src/main/java/jenkins/util/Listeners.java
Comment (timja) [inline core/src/main/java/jenkins/util/Listeners.java:78]: probably not used anywhere in the wild? it came in a very recent weekly can we just bin it?

Comment (jglick) [reply core/src/main/java/jenkins/util/Listeners.java:78]: Probably, just wanted to err on the safe side.

**Pull request merged**
